### ISSUE TITLE
Support `.markdownlint.js`

### ISSUE THREAD

**Issue by Mister-Hope** [DavidAnson/markdownlint-cli2#304]
`.markdownlint.js` should be supported and its file type (cjs/esm) should be based on current `package.json`.

This is a problem because I need to extra declare `.markdownlint.mjs` / `.markdownlint.cjs` in eslint and typescript config files in all my repo to have it included, while all other files are `js`.

Anyway its weird to have a `.cjs` file in cjs package and a `.mjs` file in esm package.

Other tools like `commitlint` `eslint` `vite`, ... all support this.

**Comment by DavidAnson** [comment]
The current behavior of ClI2 is deliberate. In my mind, the "js" extension should no longer be used because it is ambiguous.

That said, I am curious if the behavior in other tools is deliberate or accidental - and whether they had to write code to support it.

**Comment by Mister-Hope** [comment]
>  In my mind, the "js" extension should no longer be used because it is ambiguous.

Not meaning to be rude, but check your current repo plz? Most of the popular repo is written in full `.mjs` or `.cjs`, then where does the point come from?

In a `type: "commonjs/module"` package, a `.js` is the same as `.cjs/.mjs`.

**Comment by Mister-Hope** [comment]
Some Refs:

- https://eslint.org/docs/latest/use/configure/configuration-files
- https://commitlint.js.org/reference/configuration.html
- https://vitejs.dev/config/
- https://prettier.io/docs/en/configuration

**Comment by Mister-Hope** [comment]
Maybe you can use this pkg directly: https://github.com/cosmiconfig/cosmiconfig/tree/v8.2.0

**Comment by DavidAnson** [comment]
The point is that you (or an editor) cannot tell from looking at JS extension what type of file it is. What's more, copying that file from one project to another may not work because the projects are configured differently. None of that is an issue with the CJS or MJS extension.

While it is less obvious in this repository, I use CJS or MJS for new files except where existing files already have the JS extension. That means this tends to happen most in new test files and you can see that in other repositories I maintain. The new `esint.config.mjs` in `next` branch is a good example. 

I will have a look at some of the links you sent, but I do not think this behavior raises to the level of introducing an entirely new configuration system. As you note above, everything works fine today - the issue is just that you do not want to deal with different extensions.

**Comment by Mister-Hope** [comment]
Yes, everything could work fine under current package, but as all popular lint tool support `js` config file and have its type auto-inferred (see docs I providered ealier), then it could be a kind of "common feature lacking” for the package itself.

**Comment by Mister-Hope** [comment]
Tools eslint, prettier, stylelint and lint-staged support all [cosmiconfig](https://github.com/cosmiconfig/cosmiconfig) format (eslint up to v8, since v9 introduce flat config and it's a breaking change, a script config file is required)

Tools rollup, commintlint and vite are even more aggressive, they all preprocess the config file, so esm config file is even allowed with `.js` in a cjs pkg.

**Comment by Mister-Hope** [comment]
> The point is that you (or an editor) cannot tell from looking at JS extension what type of file it is. 

Solution can be found in the ref I provided

![image](https://github.com/DavidAnson/markdownlint-cli2/assets/33315834/e564ed08-c598-41fd-9fe7-def7d1bc5543)
![image](https://github.com/DavidAnson/markdownlint-cli2/assets/33315834/9e180eec-1d51-47ee-9dc0-e4c01f5cce15)
![image](https://github.com/DavidAnson/markdownlint-cli2/assets/33315834/5733d755-dbd6-49ae-86c1-4e82a0178a48)

 What's more, copying that file from one project to another may not work because the projects are configured differently. 

> I don't think this should be an issue, if a people just want to quickly re-use an existing file and the type is different between 2 projects, he just need to rename the file to have `cjs`/mjs` extension.

**Comment by Mister-Hope** [comment]
> The point is that you (or an editor) cannot tell from looking at JS extension what type of file it is. 

Solution can be found in the ref I provided

![image](https://github.com/DavidAnson/markdownlint-cli2/assets/33315834/e564ed08-c598-41fd-9fe7-def7d1bc5543)
![image](https://github.com/DavidAnson/markdownlint-cli2/assets/33315834/9e180eec-1d51-47ee-9dc0-e4c01f5cce15)
![image](https://github.com/DavidAnson/markdownlint-cli2/assets/33315834/5733d755-dbd6-49ae-86c1-4e82a0178a48)

> What's more, copying that file from one project to another may not work because the projects are configured differently. 

I don't think this should be an issue, if a people just want to quickly re-use an exisiting file and the type is different between 2 projects, he just need a rename.

**Comment by DavidAnson** [comment]
I already do not support JSON and YML extensions for the new options file format (preferring JSONC and YAML) for simplicity and I think this is similar. Except here there is the added complexity of JS meaning CJS and/or MJS depending on where the file lives.

Where we differ is that I don't think this is a "feature" because it doesn't enable any unique behavior. I see why people think it is a "simplification", but it also introduces ambiguity and complexity that does not need to exist.

I will leave this issue open for comments and to think about more.

**Comment by Likakong** [comment]
> `.markdownlint.js`គួរតែត្រូវបានគាំទ្រ ហើយប្រភេទឯកសាររបស់វា (cjs/esm) គួរតែផ្អែកលើបច្ចុប្បន្ន `package.json`.
> 
> នេះជាបញ្ហាព្រោះខ្ញុំត្រូវការបន្ថែម declare `.markdownlint.mjs`/ `.markdownlint.cjs`in eslint និង typescript config files នៅក្នុង repo ទាំងអស់របស់ខ្ញុំដើម្បីឱ្យវារួមបញ្ចូល ខណៈពេលដែលឯកសារផ្សេងទៀតទាំងអស់គឺ `js`.
> 
> អ្វីក៏ដោយវាចំលែកដែលមាន `.cjs`ឯកសារនៅក្នុងកញ្ចប់ cjs និង `.mjs`ឯកសារនៅក្នុងកញ្ចប់ esm ។
> 
> ឧបករណ៍ផ្សេងទៀតដូចជា `commitlint` `eslint` `vite`, ... ទាំងអស់គាំទ្រវា។

https://github.com/DavidAnson/markdownlint-cli2/issues/304#issue-2258062485
